server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2687,4 +2665,24 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
             }
         };
     }
+
+    /**
+     * Rollback the current engine to the safe commit, then replay local translog up to the global checkpoint.
+     */
+    void resetEngineToGlobalCheckpoint() throws IOException {
+        assert getActiveOperationsCount() == 0 : "Ongoing writes [" + getActiveOperations() + "]";
+        sync(); // persist the global checkpoint to disk
+        final long globalCheckpoint = getGlobalCheckpoint();
+        final long maxSeqNo = seqNoStats().getMaxSeqNo();
+        logger.info("resetting replica engine from max_seq_no [{}] to global checkpoint [{}]", maxSeqNo, globalCheckpoint);
+        final Engine resettingEngine;
+        synchronized (mutex) {
+            IOUtils.close(currentEngineReference.getAndSet(null));
+            resettingEngine = createNewEngine(newEngineConfig());
```

Comment:
if we do this, why did we need to change how createNewEngine behaved (i.e.,  update currentEngineReference etc.)

Reply:
Ok, I copied this change from the previous PR where we have a transition to a read-only engine. I'll revert this change and re-introduce later.